DataDog/datadog-operator
# [CECO-754][introspection] Clean up unused ds/eds

### What does this PR do?

Cleans up unused DS/EDSs. In a cluster with cos and ubuntu node pools, we start with 2 DS:
```
$ kubectl get pods
NAME                                         READY   STATUS    RESTARTS   AGE
datadog-agent-gcp-cos-bp9hj                  3/3     Running   0          5m46s
datadog-agent-gcp-cos-rp5jl                  3/3     Running   0          5m46s
datadog-agent-gcp-ubuntu-h8d95               3/3     Running   0          2m
datadog-agent-gcp-ubuntu-mkfbq               3/3     Running   0          2m
datadog-cluster-agent-54cbb867bc-jhvw5       1/1     Running   0          5m46s
operator-datadog-operator-5c965d787d-2zxh8   1/1     Running   0          10m
$ kubectl get ds
NAME                       DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
datadog-agent-gcp-cos      2         2         2       2            2           <none>          5m52s
datadog-agent-gcp-ubuntu   2         2         2       2            2           <none>          2m6s
$ kubectl get dd
NAME      AGENT             CLUSTER-AGENT     CLUSTER-CHECKS-RUNNER   AGE
datadog   Running (4/4/4)   Running (1/1/1)                           5m55s
```

After deleting the ubuntu node pool, only cos nodes are left so only the `datadog-agent-gcp-cos` DS is needed:
```
$ kubectl get ds
NAME                    DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
datadog-agent-gcp-cos   2         2         2       2            2           <none>          9m32s
$ kubectl get pods
NAME                                         READY   STATUS    RESTARTS   AGE
datadog-agent-gcp-cos-bp9hj                  3/3     Running   0          9m33s
datadog-agent-gcp-cos-rp5jl                  3/3     Running   0          9m33s
datadog-cluster-agent-54cbb867bc-jhvw5       1/1     Running   0          9m33s
operator-datadog-operator-5c965d787d-2zxh8   1/1     Running   0          14m
$ kubectl get dd
NAME      AGENT             CLUSTER-AGENT     CLUSTER-CHECKS-RUNNER   AGE
datadog   Running (2/2/2)   Running (1/1/1)                           31m
```

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Minimum Agent Versions

Are there minimum versions of the Datadog Agent and/or Cluster Agent required?

* Agent: n/a
* Cluster Agent: n/a

### Describe your test plan

1. Spin up node pools with multiple providers (e.g. cos and ubuntu) so that multiple DS/EDSs are created
2. Delete some node pools so that there are fewer providers than before
3. Ensure that DS/EDSs that do not spin up agent pods are deleted

### Checklist
